packages/fluentui/react-northstar/src/components/Tree/hooks/useTreeActiveState.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react';
+import { useAutoControlled } from '@fluentui/react-bindings';
+import { TreeProps } from '../Tree';
+import { useStableProps } from './useStableProps';
+import { FlatTree } from './flattenTree';
+import { findIndex, removeItemAtIndex } from './utils';
+import * as _ from 'lodash';
+
+export interface TreeActiveState {
+  activeItemIds: string[];
+  flatTree: FlatTree;
+  toggleActive: (ids: string[], e: React.SyntheticEvent) => void;
+}
+
+export function useTreeActiveState(
+  props: Pick<TreeProps, 'defaultActiveItemIds' | 'activeItemIds' | 'exclusive'>,
+  flatTree: FlatTree,
+): TreeActiveState {
+  const initialActiveItemIds = React.useMemo(() => {
+    const initalValue = [];
+    Object.keys(flatTree).forEach(key => {
+      if (flatTree[key].expanded) {
+        initalValue.push(key);
+      }
+    });
+    return initalValue;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []); // initialValue only needs to be computed on mount
+
+  const [activeItemIds, setActiveItemIdsState] = useAutoControlled<string[]>({
+    defaultValue: props.defaultActiveItemIds,
+    value: props.activeItemIds,
+    initialValue: initialActiveItemIds,
+  });
+
+  // compute a new flattened tree based on activeItemIds state
+  const updatedFlatTree = React.useMemo(() => {
+    const updatedFlatTree = { ...flatTree };
+    Object.keys(updatedFlatTree).forEach(key => {
+      if (updatedFlatTree[key].expanded) {
+        updatedFlatTree[key].expanded = false;
+      }
+    });
```

Comment:
You don't want to iterate the object twice
```suggestion
    const updatedFlatTree = {};
    Object.keys(flatTree).forEach(key => {
     updatedFlatTree[key] = {...flatTree[key], expanded: false};
    });
```

Reply:
`reduce` instead of `forEach` will mean you don't have `updatedFlatTree` existing as an empty object, but that's more of a nit

Reply:
I see, your way has less loop. I did a test and it is twice faster for 100 items. Going to change to use the same idea :) 
```suggestion
  const updatedFlatTree = {};
  Object.keys(flatTree).forEach((key) => {
    updatedFlatTree[key] = { ... flatTree[key] };
    if (updatedFlatTree[key].expanded) {
      updatedFlatTree[key].expanded = false;
    }
  });
```
(I didn't commit your suggestions directly because I wanted to add a check of `if (updatedFlatTree[key].expanded)`, so I keep the expanded=undefined to continue be undefined)